### PR TITLE
Add supporting APIs for evacuation

### PR DESCRIPTION
**Pull request by xyuanlu**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

#2589

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

This change add 2 APIs.
isEvacuateFinished 
    Return if instance operation 'Evacuate' is finished.
isReadyForPreparingJoiningCluster 
    Return if instance is ready for preparing joining cluster. The instance should have no current state, no pending message and tagged with operation that exclude the instance from Helix assignment.

### Tests

- [X] The following tests are written for this issue:
Updated testInstanceOperation

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

**Review by MarkGaox — APPROVED** [review]
LGTM

File: helix-core/src/main/java/org/apache/helix/HelixAdmin.java [hunk 1]
```diff
@@ -738,4 +738,15 @@ Map<String, Boolean> validateResourcesForWagedRebalance(String clusterName,
...
+  /**
+   * Return if instance operation 'Evacuate' is finished.
+   * Only return true if there is no current state on the instance and that instance is still alive.
+   * @param clusterName
+   * @param instancesNames
+   * @return
+   */
+  boolean isEvacuateFinished(String clusterName, String instancesNames);
+
+  boolean isPrepopulateReady(String clusterName, String instancesNames);
```
Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/HelixAdmin.java:751]: java doc

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java [hunk 2]
```diff
@@ -406,7 +407,41 @@ public ZNRecord update(ZNRecord currentData) {
...
+  public boolean isEvacuateFinished(String clusterName, String instanceName) {
+    HelixDataAccessor accessor =
+        new ZKHelixDataAccessor(clusterName, new ZkBaseDataAccessor<ZNRecord>(_zkClient));
+    PropertyKey.Builder keyBuilder = accessor.keyBuilder();
+
+    // check the instance is alive
+    LiveInstance liveInstance = accessor.getProperty(keyBuilder.liveInstance(instanceName));
+    if (liveInstance == null) {
+      logger.warn("instance {} in cluster {} is not alive.", instanceName, clusterName);
+      return false;
```
Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java:421]: Let's make it true as it designed for whether we can continue following operation when the evacuated is done.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java [hunk 3]
```diff
@@ -406,7 +407,41 @@ public ZNRecord update(ZNRecord currentData) {
...
+    String sessionId = liveInstance.getEphemeralOwner();
+
+    String path = PropertyPathBuilder.instanceCurrentState(clusterName, instanceName, sessionId);
+    List<String> currentStates = baseAccessor.getChildNames(path, 0);
+    if (currentStates == null) {
+      logger.warn("instance {} in cluster {} does not have live session.", instanceName,
+          clusterName);
+      return false;
+    }
+    return  currentStates.isEmpty();
```
Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java:434]: edge case... check whether has pending messages.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java [hunk 4]
```diff
@@ -406,7 +407,41 @@ public ZNRecord update(ZNRecord currentData) {
...
+  }
+
+  @Override
+  public boolean isPrepopulateReady(String clusterName, String instanceName) {
```
Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java:438]: rename it more clear.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java [hunk 5]
```diff
@@ -406,7 +407,41 @@ public ZNRecord update(ZNRecord currentData) {
...
+    return this.getInstanceConfig(clusterName, instanceName)
+        .getInstanceOperation()
+        .equals(InstanceConstants.InstanceOperation.EVACUATE.name());
```
Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java:441]: Be consistent with the operation tag. Add SWAP_OUT here.

File: helix-core/src/main/java/org/apache/helix/HelixAdmin.java (hunk 1 above)
Comment (xyuanlu) [reply helix-core/src/main/java/org/apache/helix/HelixAdmin.java]: TFTR. Updated.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java [hunk 6]
```diff
@@ -407,10 +409,60 @@ public ZNRecord update(ZNRecord currentData) {
...
+    BaseDataAccessor<ZNRecord> baseAccessor = new ZkBaseDataAccessor<ZNRecord>(_zkClient);
+    String sessionId = liveInstance.getEphemeralOwner();
+
+    String path = PropertyPathBuilder.instanceCurrentState(clusterName, instanceName, sessionId);
+    List<String> currentStates = baseAccessor.getChildNames(path, 0);
+    if (currentStates == null) {
+      logger.warn("Instance {} in cluster {} does not have live session.  The instance can be removed anyway.",
+          instanceName, clusterName);
+      return false;
+    }
```
Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java:451]: Race condition can happen if the instance is carrying over from old session current state. New session current state is empty.

**Review by junkaixue — APPROVED** [review]
Overall lgtm. Please verify the liveinstance & session carrying over logic.

**Comment by xyuanlu** [comment]
This PR is ready to be merged. Approved by @junkaixue 
Commit message: Add supporting APIs a isEvacuateFinished and isReadyForPreparingJoiningCluster for evacuation.

Failed CI test has been fixed on master branch.

**Pull request merged**